lib/spack/env/cc

```diff
@@ -58,59 +69,140 @@ parameters=(
 # Test command is used to unit test the compiler script.
 #   SPACK_TEST_COMMAND
 
-# die()
+# die MESSAGE
 # Prints a message and exits with error 1.
-function die {
+die() {
     echo "$@"
     exit 1
 }
 
-# read input parameters into proper bash arrays.
-# SYSTEM_DIRS is delimited by :
-IFS=':' read -ra SPACK_SYSTEM_DIRS <<< "${SPACK_SYSTEM_DIRS}"
+# empty VARNAME
+# whether a variable is empty
+empty() {
+    eval "test -z \"\${$1}\""
+}
+
+# setsep LISTNAME
+# set the variable 'sep' to the separator for a list with name LISTNAME
+# There are three types of lists:
+#   1. regular lists end with _list and are separated by $lsep
+#   2. directory lists end with _dirs or _DIRS and are separated by ':'
+#   3. any other list is assumed to be separated by spaces: " "
+setsep() {
+    case "$1" in
+        *_dirs|*_DIRS) sep=':'      ;;
+        *_list)        sep="$lsep"  ;;
+        *)             sep=" "      ;;
+    esac
+}
+
+# prepend LISTNAME ELEMENT [SEP]
+#
+# Prepends element to list, assuming list is separted by SEP.
+# Handles empty lists and single-element lists.
+prepend() {
+    varname="$1"
+    elt="$2"
+
+    if empty "$varname"; then
+        eval "$varname=\"\${elt}\""
+    else
+        # get appropriate separator for the list we're appending to
+        setsep "$varname"
+        eval "$varname=\"\${elt}${sep}\${$varname}\""
+    fi
+}
+
+# append LISTNAME ELEMENT [SEP]
+#
+# Appends element to list, assuming list is separted by SEP.
+# Handles empty lists and single-element lists.
+append() {
+    varname="$1"
+    elt="$2"
+
+    if empty "$varname"; then
+        eval "$varname=\"\${elt}\""
+    else
+        # get appropriate separator for the list we're appending to
+        setsep "$varname"
+        eval "$varname=\"\${$varname}${sep}\${elt}\""
+    fi
+}
+
+# extend LISTNAME1 LISTNAME2 [PREFIX]
+#
+# Append elements from second list to first.
+# If PREFIX is provided, prepend it to each element.
+extend() {
+    # figure out appropriate IFS for the list we're reading
+    setsep "$2"
+    if [ "$sep" != " " ]; then
+        IFS="$sep"
+    fi
+    eval "for elt in \${$2}; do append $1 \"$3\${elt}\"; done"
```

Comment:
This is a fun idea I've used in the past:
```bash
# with_IFS_for VARNAME EVAL-STRING
# Temporarily set IFS to the value necessary to split the list at VARNAME,
# then eval EVAL-STRING.
with_IFS_for() {
    set_IFS "$1"
    eval "$2"
    rc="$?"
    unset IFS
    return "$rc"
}
```

It's just complex enough to not be worth using here imho, but I thought I'd bring it up. Can also do `shift` then execute `"$@"` instead of `eval` specifically.